image_to_table/image_with_word_boxes.py

```diff
@@ -1,146 +1,87 @@
-    ]
-    return column_boxes
-
-
-def show_all_boxes_intersecting(filename, row_boxes, column_boxes):
-    for row_box in row_boxes:
-        for column_box in column_boxes:
-            show_image_with_word_boxes(filename, [row_box, column_box])
-
-
-def get_row(row_box, column_boxes, original_boxes):
-    new_row = [get_cell_in_row(column_box, original_boxes, row_box) for column_box in column_boxes]
-    return new_row
-
-
-def get_cell_in_row(column_box, original_boxes, row_box):
-    row_column = []
-    for i, original_box in enumerate(original_boxes):
-        if original_box.is_inside_box(column_box) and original_box.is_inside_box(row_box):
-            row_column.append(original_box.text)
-    return " ".join(row_column)
-
-
-def extract_table_from_image(filename):
+
+def merge_sorted_text_boxes(boxes: List[TextBox]) -> TextBox:
+    bounding_rect = boxes[0].rect or boxes[-1].rect
+    description = " ".join(map(lambda x: x.text, boxes))
+
+    return TextBox(description, bounding_rect)
+
+
+def extract_table_from_image(filename: str, placement: List[int]) -> List[List[TextBox]]:
     image = cv2.imread(filename)
     height, width, _ = image.shape
     original_boxes = detect_text(filename)
     boxes = original_boxes
-    for box in original_boxes:
-        box.fill("black")
-    x_counts = [len(boxes_in_x(boxes, x)) for x in range(width)]
-    boxes_in_ys = [boxes_in_y(boxes, y) for y in range(height)]
-    y_counts = [len(l) for l in boxes_in_ys]
-    row_boxes = make_row_boxes(y_counts, width)
-    column_boxes = make_column_boxes(x_counts, height)
-    """
-    for box in row_boxes + column_boxes:
-        box.plot(color="black")
-    """
-    plt.axis("off")
-    plt.savefig("table_b.png")
-    plt.show()
 
-    table = [get_row(row_box, column_boxes, original_boxes) for row_box in row_boxes]
+    # Sort boxes by row
+    sorted_boxes = sorted(boxes, key=lambda box: box.rect.y)
+
+    rows = merge_into_rows(sorted_boxes)
+    table = merge_into_columns(rows, placement)
 
     return table
+
+
+def merge_into_rows(boxes: List[TextBox], max_distance: int = 5) -> List[List[TextBox]]:
+    """Merges text boxes with similar heights into rows
+
+    :param max_distance: Maximum height difference between rows to be considered the same row.
+    """
+    ys = np.asarray(list(map(lambda x: x.rect.y, boxes)))
+    diffs = np.diff(ys)
+    indices = np.where(diffs > max_distance)[0] + 1
+
+    rows = [boxes[: indices[0]]]
+    for i in range(len(indices) - 1):
+        rows.append(boxes[indices[i] : indices[i + 1]])
+
+    return rows
+
+
+def merge_into_columns(rows: List[List[TextBox]], placements: List[int]) -> List[List[TextBox]]:
+    bulks = []
+    for row in rows:
+        bulk = []
+        row = sorted(row, key=lambda x: x.rect.tl.x)
+        tr_xs = list(map(lambda x: x.rect.tr.x, row))
+
+        last_index = 0
+        for placement in placements:
+            index = bisect.bisect_left(tr_xs, placement, lo=last_index)
```

Comment:
bisect! nice!

Reply:
Never used it before, but thought it would be a waste to search the whole thing when the list is already sorted. This is what I found 😅

Reply:
I'm guessing it does binary search, which is definitely much better than linear search!